File: Day238/reconstructItinerary.java
package Day238;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;


public class reconstructItinerary {
    public static void dfs(String s, Map<String, PriorityQueue<String>> map, LinkedList<String> list) {
        PriorityQueue<String> temp = map.get(s);
        while (temp!= null && !temp.isEmpty()) {
            dfs(temp.poll(), map, list);
        }
        list.addFirst(s);
    }
    public static List<String> findItinerary(List<List<String>> tickets) {
        LinkedList<String> answer = new LinkedList<>();
        Map<String, PriorityQueue<String>> map = new HashMap<>();
        for (int i=0; i<tickets.size(); i++) {
            if (!map.containsKey(tickets.get(i).get(0))) {
                map.put(tickets.get(i).get(0), new PriorityQueue<>());
            }
            map.get(tickets.get(i).get(0)).add(tickets.get(i).get(1));
        }
        System.out.println(map);
        dfs("JFK", map, answer);
        return answer;
    }
    public static void main(String[] args) {
        List<List<String>> tickets = new ArrayList<>();
        String arr[][] ={{"JFK","KUL"},{"JFK","NRT"},{"NRT","JFK"}};
        for (int i=0; i<arr.length; i++) {
            List<String> temp = new ArrayList<>();
            for (int j=0; j<arr[i].length; j++) {
                temp.add(arr[i][j]);
            }
            tickets.add(temp);
        }
        
        // System.out.println(tickets);
        System.out.println(findItinerary(tickets));
    }   
}

File: Day238/reconstructItinerary.go
package main

import (
	"fmt"
	"sort"
)

func findItinerary(tickets [][]string) []string {
    adj := make(map[string][]string)
    for _, i := range tickets {
        adj[i[0]] = append(adj[i[0]], i[1])
    }
    for key:= range adj {
        sort.Sort(sort.Reverse(sort.StringSlice(adj[key])))
    }
    answer := make([]string, 0)
    var dfs func(source string)
    dfs = func(source string) {
        for len(adj[source]) > 0 {
            to := adj[source][len(adj[source])-1]
            adj[source] = adj[source][:len(adj[source])-1]
            dfs(to)
        }
        answer = append(answer, source)
    }
    dfs("JFK")
    for i:=0 ; i<len(answer)/2; i++ {
        answer[i], answer[len(answer)-i-1] = answer[len(answer)-i-1], answer[i]
    }
    return answer
}
func main () {
	tickets := [][]string{{"JFK","KUL"},{"JFK","NRT"},{"NRT","JFK"}}
	fmt.Println(findItinerary(tickets))
}